src/MongoDB.Analyzer/Core/Builders/BuildersVariablesResolver.cs

```diff
@@ -0,0 +1,390 @@
+﻿// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+
+namespace MongoDB.Analyzer.Core.Builders;
+
+internal static class BuildersVariablesResolver
+{
+    private sealed record VariableValue(
+        ExpressionAnalysisContext ExpressionAnalysisContext,
+        List<Location> Locations);
+
+    private sealed record ContextInformation(
+        SemanticModel SemanticModel,
+        List<VariableValue> ProcessedVariables,
+        Dictionary<string, VariableValue> CurrentVariablesValues,
+        HashSet<SyntaxNode> BuilderNodesProcessed,
+        Dictionary<SyntaxNode, ExpressionAnalysisContext> BuilderToAnalysisContext);
+
+    private sealed record EvaluationResult(
+        bool CanEvaluate,
+        Dictionary<SyntaxNode, VariableValue> SyntaxNodeToVariableValueMap);
+
+    public static List<ExpressionAnalysisContext> ResolveVariables(
+        Dictionary<SyntaxNode, ExpressionAnalysisContext> builderToAnalysisContextMap,
+        SemanticModel semanticModel)
+    {
+        var expressionAnalysisContexts = new List<ExpressionAnalysisContext>();
+        var variableValuesList = new List<VariableValue>();
+        var builderNodesProcessed = new HashSet<SyntaxNode>();
+        ContextInformation contextInformation = new ContextInformation(semanticModel, variableValuesList, new Dictionary<string, VariableValue>(), builderNodesProcessed, builderToAnalysisContextMap);
+        ProcessTree(semanticModel.SyntaxTree.GetRoot(), contextInformation);
+
+        var newAnalysisContexts = variableValuesList.Select(v => v.ExpressionAnalysisContext with { Node = v.ExpressionAnalysisContext.Node with { Locations = v.Locations.ToArray() } });
```

Comment:
Let's check somehow that at least one new context was added (variableValuesList.Any() ?), and only then construct new contexts.

something like:
```
var result = analysisContexts; // add parameter to ResolveVariables
if (variableValuesList.Any())
{
   
 result = variableValuesList.Select(...).ToList().
 result .AddRange(previousAnalysisContexts);
}
```

Reply:
Done